src/search.coffee

```diff
@@ -6,234 +6,257 @@ url = require 'url'
-      .first()
-      .pairs()
-      .filter ([key, value]) ->
-        # Eg. WebTotal, ImageTotal, ...
-        keyRe.test key
-      .map ([key, value]) ->
-        # Eg. WebTotal => web
-        key = keyRe.exec(key)[1].toLowerCase()
-        value = Number value
-
-        switch key
-          when 'spellingsuggestions' then ['spelling', value]
-          else [key, value]
-      .object()
-      .value()
-
-  verticalSearch: (vertical, verticalResultParser, query, options, callback) ->
-    [callback, options] = [options, {}] if _.compact(arguments).length is 4
-
-    @parallelSearch vertical, _.extend({}, options, {query}), (err, result) ->
-      return callback err if err
-      callback null, verticalResultParser result
-
-  mapResults: (results, fn) ->
-    _.chain(results)
-      .pluck('d')
-      .pluck('results')
-      .flatten()
-      .map fn
-      .value()
+  _parallelSearch: (options, callback) ->
+    allRequestOptions = []
+
+    top = Search.MAX_RESULTS
+    if options.top?
+      top = Number options.top
+      options = _.omit options, 'top'
+
+    allRequestOptions.push _.defaults {
+      count: Math.min Search.PAGE_SIZE, top - offset
+      offset
+    }, options for offset in [(options.offset || 0)...top] by Search.PAGE_SIZE
+
+    async.mapLimit allRequestOptions, @parallelLimit, _.bind(@_executeSearch, this),
+      (err, responses) ->
+        callback err if err?
+
+        data = {
+          count: 0
+          results: []
+        }
+
+        # Avoid duplicates by checking result IDs.
+        existingIds = []
+        responses.forEach (response) ->
+          data.count = response.count if response.count?
+          response.results.forEach (result) ->
+            unless result.id in existingIds
+              data.results.push result
+              existingIds.push result.id
+
+        callback null, data
+
+  counts: (query, options, callback) ->
+    [callback, options] = [options, {}] if _.compact(arguments).length is 2
+
+    sources = ['web', 'images', 'videos', 'news']
+    methods = {
+      web: _.bind @_rawWeb, this
+      images: _.bind @_rawImages, this
+      videos: _.bind @_rawVideos, this
+      news: _.bind @_rawNews, this
+    }
+
+    executeSearch = (source, callback) =>
+      # We only need to one result (zero is impossible) from most verticals.
+      # With a web search, however, the `totalEstimatedMatches` needs to be
+      # checked from a higher page for accurate data.
+      #
+      # The 1,000 value comes from empirical data. It seems that after 600
+      # results, the accuracy gets quite consistent and accurate. I picked 1,000
+      # just to be in the clear. It also doesn't matter if there are fewer than
+      # 1,000 results.
+      pagination = {offset: 0, top: 1}
+      if source is 'web'
+        _.each pagination, (value, key) ->
+          pagination[key] += 1000
+
+      methods[source] query, (_.defaults pagination, options), callback
```

Comment:
I see what you're doing here. Prefer this instead:

`_.defaults(pagination, options)`